### PR TITLE
Add in Challenge service

### PR DESCRIPTION
**Pull request by thomasyu888**
* add in `Challenge` response object
* add in `ChallengeApi` class
* Move `synapseservices` into `challengeutils`
* commandline tool - list registered challenges

### CODE REVIEW

**Comment by thomasyu888** [comment]
#110

**Comment by thomasyu888** [comment]
After discussion with jordan:

1. For a library that is consumed by others it's nice to have a small dependency list since you don't know how people will use your package. If they aren't installing it into a venv then any additional dependency could conflict with another lib in the user's system python. ( Don't use dataclass and dataclass_json)

2. split up the api calls from the data class.

File: challengeutils/challenge.py [hunk 1]
```diff
@@ -1,8 +1,173 @@
...
+
+class ChallengeApi:
+    """Challenge services
+    https://docs.synapse.org/rest/index.html#org.sagebionetworks.repo.web.controller.ChallengeController
+
+    Args:
+        id: The ID of this Challenge object
+        project: synapseclient.Project or its id
+        team: synapseclient.Team or its id
+
```
Comment (vpchung) [inline challengeutils/challenge.py:23]: We should probably keep the whitespace consistent, e.g. some docstrings have a blank line before the closing triple-quotes while others don't.

File: challengeutils/challenge.py [hunk 2]
```diff
@@ -1,8 +1,173 @@
...
+        challenge = self.syn.restPUT(f'/challenge/{challengeid}',
+                                     str(new_challenge))
+        return Challenge(**challenge)
+
+    def delete_challenge(self, challengeid: str):
+        """Deletes a Synapse Challenge
+
+        Args:
+            challengeid: A Synapse Challenge id
+        """
```
Comment (vpchung) [inline challengeutils/challenge.py:91]: See above.

File: challengeutils/synapseservices/base_service.py [hunk 3]
```diff
@@ -172,21 +208,18 @@ def to_dict(self):
...
 
-    def __eq__(self, other):
+    def __eq__(self, other) -> bool:
         """Returns true if both objects are equal"""
         return self.__dict__ == other.__dict__
 
-    def __ne__(self, other):
+    def __ne__(self, other) -> bool:
         """Returns true if both objects are not equal"""
         return not self == other
```
Comment (vpchung) [inline challengeutils/synapseservices/base_service.py:225]: Interesting... wouldn't `return self != other` be more readable...?

File: challengeutils/synapseservices/challenge.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    def id(self, id):
+        """Sets the id of this Activity.
+        :param id: The id of this Activity.
+        :type id: str
+        """
+        self._id = id
+
+    @property
+    def projectId(self):
+        """Gets the id of this Activity.
```
Comment (vpchung) [inline challengeutils/synapseservices/challenge.py:57]: I think this docstring needs to be updated? It's returning the projectId of the Activity right, not id?

File: challengeutils/synapseservices/challenge.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+        """
+        return self._etag
+
+    @etag.setter
+    def etag(self, etag):
+        """Sets the id of this Activity.
+        :param id: The id of this Activity.
+        :type id: str
+        """
+        self._id = etag
```
Comment (vpchung) [inline challengeutils/synapseservices/challenge.py:85]: `self._etag = etag`?

File: challengeutils/synapseservices/challenge.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+                 str = None, participantTeamId: str = None):
+
+        self.openapi_types = {
+            'id': str,
+            'projectId': str,
+            'etag': str,
+            'participantTeamId': str
+        }
+
+        self.attribute_map = {
```
Comment (vpchung) [inline challengeutils/synapseservices/challenge.py:27]: I don't think I understand the purpose of this...

File: challengeutils/challenge.py (hunk 1 above)
Comment (thomasyu888) [reply challengeutils/challenge.py:23]: Agreed, thanks.  It is actually pep8 standards to have a blank line before the closing.

File: challengeutils/synapseservices/base_service.py (hunk 3 above)
Comment (thomasyu888) [reply challengeutils/synapseservices/base_service.py:225]: See this post: https://stackoverflow.com/questions/4352244/python-should-i-implement-ne-operator-based-on-eq.  Seems that I should actually remove this for python3.  This was legacy python2 code.

File: challengeutils/challenge.py (hunk 1 above)
Comment (vpchung) [reply challengeutils/challenge.py:23]: Oh, I thought it was blank line after the closing?

Comment (thomasyu888) [reply challengeutils/challenge.py:23]: It appears we are both wrong: https://google.github.io/styleguide/pyguide.html

File: challengeutils/synapseservices/challenge.py (hunk 6 above)
Comment (thomasyu888) [reply challengeutils/synapseservices/challenge.py:27]: You can see the purpose of this in the `base_service.py`.  Essentially a lot of API's use camelCase, but a lot of python now uses snakecase.
```
        attribute_map: The key is attribute name and the
                       value is json key in definition.
```

I originally had:
```
        self.attribute_map = {
            'id': 'id',
            'projectid': 'projectId',
            'etag': 'etag',
            'participant_teamid': 'participantTeamId'
        }
```

Which then wouldn't throw the pep8 warning:
```
Attribute name "projectId" doesn't conform to snake_case naming style
```

That being said, I still wanted people to be able to create the class with the response object, so I stuck with camelCase.

Comment (vpchung) [reply challengeutils/synapseservices/challenge.py:28]: Oh I see now. Got it :thumbsup:

File: challengeutils/challenge.py [hunk 7]
```diff
@@ -1,8 +1,218 @@
...
+        syn: Synapse connection
+        userid: Specify userid if you want to know the challenges
+                another Synapse user is registered to.
+
+    Yields:
+        A synapseclient.Project
+
+    """
+    challenge_api = ChallengeApi(syn=syn)
+    # This will return the logged in user profile is None is passed in
```
Comment (thomasyu888) [inline challengeutils/challenge.py:169]:
```suggestion
    # This will return the logged in user profile if None is passed in
```

**Pull request merged**